### ISSUE TITLE
Add a Serializer mode that automatically match class with its performance best Serializer 

### ISSUE THREAD

**Issue by joeyleeeeeee97** [EsotericSoftware/kryo#809]
Now Kryo provides several serializers with different performance, users need to make sure their class layout match specific requirements to achieve a certain performance level.

In the product environment,  it is common to add fields to Class. Currently, I have to manually identify few unmatched types and customize their serializer. ( I don't want to globally reset to a more compatible because most of my Classes still are totally the same, I want to get better performance).

I wonder if this is a common issue and whether we could provide a default serializer mode for that.

Some rough ideas on how we could achieve that:
1. Write the entire class layout on first serialize and associate it with an Id
2. Deserialize check if the Id specified layout match the local layout
3. If yes in 2,  automatically associate it with FieldSerializer
4. If no in 2, associate it with CompatibleFieldSerializer

I'm just brainstorming here.

**Comment by theigl** [comment]
`CompatibleFieldSerializer` is an extension of `FieldSerializer`. The main difference is that the former writes the entire class layout on first serialization and the value class of each field so it can be read even if the field is removed.

I'm assuming that most of the additional overhead of `CompatibleFieldSerializer` is the result of writing the layout. If that is the case, your suggestion would not improve performance.

Your suggested serializer would also have to write the class of each field, because otherwise `CompatibleFieldSerializer` would not support removing fields from a class.

So basically, you would end up with all the code in `CompatibleFieldSerializer` again.

**Comment by theigl** [comment]
`CompatibleFieldSerializer` is an extension of `FieldSerializer`. The main difference is that the former writes the entire class layout on first serialization and the value class of each field so it can be read even if the field is removed.

I'm assuming that most of the additional overhead of `CompatibleFieldSerializer` is the result of writing the layout. If that is the case, your suggestion would not improve performance.

Your suggested serializer would also have to write the class of each field, because otherwise `CompatibleFieldSerializer` would not support removing fields from a class.

So basically, you would end up with all the code in `CompatibleFieldSerializer` again.

**Comment by theigl** [comment]
I might be overlooking something though. If you feel your approach might improve performance, I'm open for a PR.

**Comment by joeyleeeeeee97** [comment]
@theigl  Yes, you're right. I ran some JMH tests and found CompatibleFieldSerializer and  FieldSerializer similar performance if we don't count the cost of writing metadata.

**Comment by theigl** [comment]
@joeyleeeeeee97: Thanks for verifying! Closing this issue for now.
